### DISCUSSION TITLE
Looping by valid time with METplus TCPairs

### DISCUSSION THREAD

**Discussion by mollybsmith-noaa** [dtcenter/METplus#985] [Configuration] [answered]
Is it possible to loop by valid time when verifying with METplus's TCPairs? I have a workflow that succeeds when looping by init time, with these variables:

PROCESS_LIST = TCPairs
LOOP_BY = INIT

INIT_TIME_FMT = %Y%m%d%H
INIT_BEG = 2021070100
INIT_END = 2021070100

INIT_INCREMENT = 21600 

But when I replace every instance of INIT with VALID, I get errors related to missing INIT fields, such as configparser.NoOptionError: No option 'INIT_TIME_FMT' in section: 'config'. This is despite the fact that the LOOP_BY variable is definitely set to VALID.

Is there a correct way to do this? With realtime verification, verification by valid time really has to be used instead of init time, or there's no way to verify the forecast lead times that haven't happened yet.

Thanks!

**Comment by georgemccabe** [comment, marked as answer]
Hi Molly,

The short answer is yes, you can loop by valid time with TCPairs. You will have to set:

`LOOP_ORDER = times`

In METplus 4.0.0 you will have to keep INIT_TIME_FMT and INIT_BEG set in your config file even though they won't be used. I'm actually currently updating the wrapper now to fix this while making other modifications for #931, so this will not be necessary in future versions.

The caveat is when you set the loop order to "times" you can't include an analysis tool like SeriesAnalysis in the same configuration. This is because SeriesAnalysis processes a range of times and loop-by-time will run all of the processes in the PROCESS_LIST for a given runtime then increment to the next runtime. When LOOP_ORDER = processes, TCPairs wrapper runs once using the INIT time info.

I will modify the logic so that the VALID time info variables will be used if LOOP_BY = VALID and LOOP_ORDER = processes. I will create a GitHub issue for this change so that it is isolated from the work done for #931.

**Reply by mollybsmith-noaa** [reply]
Thanks, George! I will set LOOP_ORDER to times and see how that goes.

**Reply by georgemccabe** [reply]
I created a GitHub issue from this discussion (#986) and there is an open pull request to review the changes. With these changes, you will be able to loop by valid time without having to set the INIT_ variables.
